src/java.base/share/classes/jdk/internal/ref/CleanerImpl.java

```diff
@@ -231,4 +223,135 @@ protected void performCleanup() {
+         * Capacity for a single node in the list.
+         * This balances memory overheads vs locality vs GC walking costs.
+         */
+        static final int NODE_CAPACITY = 4096;
+
+        /**
+         * Head node. This is the only node where PhantomCleanables are
+         * added to or removed from. This is the only node with variable size,
+         * all other nodes linked from the head are always at full capacity.
+         */
+        private Node head;
+
+        /**
+         * Cached node instance to provide better behavior near NODE_CAPACITY
+         * threshold: if list size flips around NODE_CAPACITY, it would reuse
+         * the cached node instead of wasting and re-allocating a new node all
+         * the time.
+         */
+        private Node cache;
+
+        public CleanableList() {
+            reset();
+        }
+
+        /**
+         * Testing support: reset list to initial state.
+         */
+        synchronized void reset() {
+            this.head = new Node();
+        }
+
+        /**
+         * Returns true if cleanable list is empty.
+         *
+         * @return true if the list is empty
+         */
+        public synchronized boolean isEmpty() {
+            // Head node size is zero only when the entire list is empty.
+            return head.size == 0;
+        }
+
+        /**
+         * Insert this PhantomCleanable in the list.
+         */
+        public synchronized void insert(PhantomCleanable<?> phc) {
+            if (head.size == NODE_CAPACITY) {
+                // Head node is full, insert new one.
+                // If possible, pick a pre-allocated node from cache.
+                Node newHead;
+                if (cache != null) {
+                    newHead = cache;
+                    cache = null;
+                } else {
+                    newHead = new Node();
+                }
+                newHead.next = head;
+                head = newHead;
+            }
+            assert head.size < NODE_CAPACITY;
+
+            // Put the incoming object in head node and record indexes.
+            phc.node = head;
+            phc.index = head.size;
+            head.arr[head.size] = phc;
+            head.size++;
+        }
+
+        /**
+         * Remove this PhantomCleanable from the list.
+         *
+         * @return true if Cleanable was removed or false if not because
+         * it had already been removed before
+         */
+        public synchronized boolean remove(PhantomCleanable<?> phc) {
+            if (phc.node == null) {
+                // Not in the list.
+                return false;
+            }
+            assert phc.node.arr[phc.index] == phc;
+            assert head.size > 0;
+
+            // Replace with another element from the head node, as long
+            // as it is not the same element. This keeps all non-head
+            // nodes at full capacity.
+            if (head != phc.node || (phc.index != head.size - 1)) {
```

Comment:
Creating a local final variable with the value of `head.size - 1` early in this method, and use that throughout (since `head.size - 1` is mentioned in multiple places (as indexing as well as assignment via head.size--) would make it a stable point in the logic.

Reply:
I did this local for `head.size - 1` in one of my patches, but on balance thought it does not gain much clarity. I reinstated it now...